ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule/SinglePeer.hs

```diff
@@ -0,0 +1,324 @@
+    intersperseTrunkFragments (i:is) (IsBranch:isTrunks) = Just i : intersperseTrunkFragments is isTrunks
+    intersperseTrunkFragments _ [] = error "intersperseTrunkFragments: not enough isTrunk flags"
+    intersperseTrunkFragments [] _ = error "intersperseTrunkFragments: not enough intersections"
+
+rawPeerScheduleFromTipPoints
+  :: R.StatefulGen g m
+  => g
+  -> PeerScheduleParams
+  -> (b -> SlotNo)
+  -> [(IsTrunk, [Int])]
+  -> Vector b
+  -> [Vector b]
+  -> [Maybe Int]
+  -> m ([(DiffTime, b)], [(DiffTime, b)], [(DiffTime, b)])
+rawPeerScheduleFromTipPoints g psp slotOfB tipPoints trunk0v branches0v intersections = do
+    let (isTrunks, tpIxs) = unzip tipPoints
+        pairedVectors = pairVectorsWithChunks trunk0v branches0v isTrunks
+        tipPointBlks = concat $ zipWith indicesToBlocks pairedVectors tpIxs
+        tipPointSlots = map slotOfB tipPointBlks
+    -- generate the tip point schedule
+    ts <- tipPointSchedule g (pspSlotLength psp) (pspTipDelayInterval psp) tipPointSlots
+    -- generate the header point schedule
+    let tpSchedules = zipMany ts tpIxs
+    hpss <- headerPointSchedule g (pspHeaderDelayInterval psp) $ zip intersections tpSchedules
+    -- generate the block point schedule
+    let (hpsPerBranch, vs) = unzip $ filter (not . null . snd .fst) $ concat
+          [ [((Nothing, hpsTrunk hps), trunk0v), ((mi, hpsBranch hps), v)]
+          | (mi, hps, v) <- zip3 intersections hpss pairedVectors
+          ]
+    bpss <- headerPointSchedule g (pspBlockDelayInterval psp) hpsPerBranch
+    let tipPointTips = zip ts tipPointBlks
+        hpsHeaders = concat $ zipWith (scheduleIndicesToBlocks trunk0v) pairedVectors hpss
+        bpsBlks = concat $ zipWith (scheduleIndicesToBlocks trunk0v) vs bpss
+    pure (tipPointTips, hpsHeaders, bpsBlks)
+
+  where
+    pairVectorsWithChunks
+      :: Vector b
+      -> [Vector b]
+      -> [IsTrunk]
+      -> [Vector b]
+    pairVectorsWithChunks trunk branches =
+       snd . mapAccumL pairVectors branches
+      where
+        pairVectors brs IsTrunk       = (brs, trunk)
+        pairVectors (br:brs) IsBranch = (brs, br)
+        pairVectors [] IsBranch       = error "not enough branches"
+
+    -- | Replaces block indices with the actual blocks
+    scheduleIndicesToBlocks :: Vector b -> Vector b -> HeaderPointSchedule -> [(DiffTime, b)]
+    scheduleIndicesToBlocks trunk v hps =
+        map (second (trunk Vector.!)) (hpsTrunk hps)
+          ++ map (second (v Vector.!)) (hpsBranch hps)
+
+    indicesToBlocks :: Vector b -> [Int] -> [b]
+    indicesToBlocks v ixs = map (v Vector.!) ixs
+
+
+-- | Merge two sorted lists.
+--
+-- PRECONDITION: The lists are sorted.
+--
+mergeOn :: Ord b => (a -> b) -> [a] -> [a] -> [a]
+mergeOn _f [] ys = ys
+mergeOn _f xs [] = xs
+mergeOn f xxs@(x:xs) yys@(y:ys) =
+    if f x <= f y
+      then x : mergeOn f xs yys
+      else y : mergeOn f xxs ys
+
+zipMany :: [a] -> [[b]] -> [[(a, b)]]
+zipMany xs0 = snd . mapAccumL (go []) xs0
```

Comment:
should you assert that the final `acc` is `[]`?

Reply:
Do you mean `fst . mapAccumL (go []) xs0`? Maybe.